dali/pipeline/operators/decoder/nvjpeg_allocator.h

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_DECODER_NVJPEG_ALLOCATOR_H_
+#define DALI_PIPELINE_OPERATORS_DECODER_NVJPEG_ALLOCATOR_H_
+
+
+#include <cuda_runtime_api.h>
+
+#include <memory>
+#include <mutex>
+#include <numeric>
+#include <string>
+#include <utility>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "dali/common.h"
+#include "dali/error_handling.h"
+#include "dali/util/cuda_utils.h"
+
+namespace dali {
+
+namespace memory {
+
+using PinnedMemoryPtr = void*;
+
+class BasicPinnedAllocator {
+ public:
+  static void PreallocateBuffers(size_t element_size_hint, size_t n_elements_hint) {
+    std::lock_guard<std::mutex> l(mutex_);
+
+    if (element_size_hint_ == 0) {
+      element_size_hint_ = element_size_hint;
+    } else {
+      DALI_ENFORCE(element_size_hint_ == element_size_hint,
+        "All instances of nvJPEGDecoder should have the same host_memory_padding.");
+    }
+    free_buffers_pool_.reserve(free_buffers_pool_.size() + n_elements_hint);
+    for (size_t i = 0; i < n_elements_hint; ++i) {
+      void* buffer;
+      CUDA_CALL(cudaHostAlloc(&buffer, element_size_hint, 0));
+      free_buffers_pool_.push_back(buffer);
+    }
+  }
+
+  static void FreeBuffers() {
+    std::lock_guard<std::mutex> l(mutex_);
+    for (auto buff : free_buffers_pool_) {
+      CUDA_CALL(cudaFreeHost(buff));
+    }
+  }
+
+  static int Alloc(void** ptr, size_t size, unsigned int flags) {
```

Comment:
I know, that this is compliant with "standard" API, but wouldn't `void &* ptr` better reflect, what's happening here? i.e. setting `ptr`?

Reply:
This is the API prototype expected by nvJPEG `nvjpegPinnedAllocator_t`

Reply:
And it'a C API